In [3]:
from tensorflow.keras.datasets import mnist #Библиотека с базой Mnist
from tensorflow.keras.models import Sequential # Подлючаем класс создания модели Sequential
from tensorflow.keras.layers import Dense # Подключаем класс Dense - полносвязный слой
from tensorflow.keras.optimizers import Adam # Подключаем оптимизатор Adam
from tensorflow.keras import utils #Утилиты для to_categorical
from tensorflow.keras.preprocessing import image #Для отрисовки изображения
import numpy as np # Подключаем библиотеку numpy
import pylab # Модуль для построения графиков
from mpl_toolkits.mplot3d import Axes3D # Модуль для трехмерной графики
#from google.colab import files #Для загрузки своей картинки
import matplotlib.pyplot as plt #Отрисовка изображений
from PIL import Image #Отрисовка изображений
#Отрисовывать изображения в ноутбуке, а не в консоль или файл
%matplotlib inline 

In [4]:
(xtrainOrg, ytrainOrg), (xtestOrg, ytestOrg)= mnist.load_data()

11493376/11490434 [==============================] - 5s 0us/step


In [5]:
xTrain=xtrainOrg.reshape(60000,784)
xTest=xtestOrg.reshape(10000,784)
print(xtrainOrg.shape,xTrain.shape,xtestOrg.shape,xTest.shape)

(60000, 28, 28) (60000, 784) (10000, 28, 28) (10000, 784)


In [6]:
xTrain = xTrain.astype('float32') 
xTrain = xTrain / 255 
xTest = xTest.astype('float32')
xTest = xTest / 255
print(xTrain.dtype,xTest.dtype)

float32 float32


In [7]:
yTrain = utils.to_categorical(ytrainOrg, 10)
yTest = utils.to_categorical(ytestOrg, 10)

In [8]:
mymodel=Sequential()
mymodel.add(Dense(10, input_dim=784, activation="relu"))
mymodel.add(Dense(400, activation="relu"))
mymodel.add(Dense(10, activation="softmax"))

In [9]:
mymodel.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])
print(mymodel.summary())

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 10)                7850      
_________________________________________________________________
dense_1 (Dense)              (None, 400)               4400      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                4010      
Total params: 16,260
Trainable params: 16,260
Non-trainable params: 0
_________________________________________________________________
None


In [10]:
mymodel.fit(xTrain, yTrain, batch_size=128, epochs=15, verbose=1,validation_split=0.1)

Epoch 1/15
422/422 [==============================] - 5s 3ms/step - loss: 0.5511 - accuracy: 0.8433 - val_loss: 0.2345 - val_accuracy: 0.9310
Epoch 2/15
422/422 [==============================] - 1s 2ms/step - loss: 0.2695 - accuracy: 0.9213 - val_loss: 0.1942 - val_accuracy: 0.9427
Epoch 3/15
422/422 [==============================] - 1s 2ms/step - loss: 0.2227 - accuracy: 0.9342 - val_loss: 0.1754 - val_accuracy: 0.9495
Epoch 4/15
422/422 [==============================] - 1s 2ms/step - loss: 0.1951 - accuracy: 0.9410 - val_loss: 0.1569 - val_accuracy: 0.9543
Epoch 5/15
422/422 [==============================] - 1s 2ms/step - loss: 0.1774 - accuracy: 0.9466 - val_loss: 0.1485 - val_accuracy: 0.9568
Epoch 6/15
422/422 [==============================] - 1s 2ms/step - loss: 0.1632 - accuracy: 0.9508 - val_loss: 0.1443 - val_accuracy: 0.9573
Epoch 7/15
422/422 [==============================] - 1s 2ms/step - loss: 0.1512 - accuracy: 0.9542 - val_loss: 0.1396 - val_accuracy: 0.9593
Epoch 

In [11]:
mymodel2=Sequential()
mymodel2.add(Dense(10, input_dim=784, activation="relu"))
mymodel2.add(Dense(400, activation="relu"))
mymodel2.add(Dense(10, activation="softmax"))

In [12]:
mymodel2.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])
print(mymodel.summary())

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 10)                7850      
_________________________________________________________________
dense_1 (Dense)              (None, 400)               4400      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                4010      
Total params: 16,260
Trainable params: 16,260
Non-trainable params: 0
_________________________________________________________________
None


In [ ]:
mymodel2.fit(xTrain, yTrain, batch_size=128, epochs=15, verbose=1,validation_split=0.5)

Epoch 1/15
235/235 [==============================] - 1s 4ms/step - loss: 1.2081 - accuracy: 0.6213 - val_loss: 0.3538 - val_accuracy: 0.8944
Epoch 2/15
235/235 [==============================] - 1s 4ms/step - loss: 0.3220 - accuracy: 0.9071 - val_loss: 0.2975 - val_accuracy: 0.9099
Epoch 3/15
235/235 [==============================] - 1s 4ms/step - loss: 0.2726 - accuracy: 0.9194 - val_loss: 0.2639 - val_accuracy: 0.9206
Epoch 4/15
235/235 [==============================] - 1s 4ms/step - loss: 0.2321 - accuracy: 0.9329 - val_loss: 0.2495 - val_accuracy: 0.9251
Epoch 5/15
235/235 [==============================] - 1s 4ms/step - loss: 0.2082 - accuracy: 0.9376 - val_loss: 0.2351 - val_accuracy: 0.9286
Epoch 6/15
235/235 [==============================] - 1s 4ms/step - loss: 0.1966 - accuracy: 0.9416 - val_loss: 0.2292 - val_accuracy: 0.9307
Epoch 7/15
235/235 [==============================] - 1s 4ms/step - loss: 0.1796 - accuracy: 0.9458 - val_loss: 0.2220 - val_accuracy: 0.9346
Epoch 

In [ ]:
mymodel3=Sequential()
mymodel3.add(Dense(10, input_dim=784, activation="relu"))
mymodel3.add(Dense(400, activation="relu"))
mymodel3.add(Dense(10, activation="softmax"))

In [ ]:
mymodel3.compile(loss="categorical_crossentropy", optimizer="adam", metrics=["accuracy"])
print(mymodel.summary())

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 10)                7850      
_________________________________________________________________
dense_1 (Dense)              (None, 400)               4400      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                4010      
Total params: 16,260
Trainable params: 16,260
Non-trainable params: 0
_________________________________________________________________
None


In [ ]:
mymodel3.fit(xTrain, yTrain, batch_size=128, epochs=15, verbose=1,validation_split=0.9)

Epoch 1/15
47/47 [==============================] - 2s 26ms/step - loss: 2.0373 - accuracy: 0.3140 - val_loss: 0.9496 - val_accuracy: 0.7287
Epoch 2/15
47/47 [==============================] - 1s 14ms/step - loss: 0.7415 - accuracy: 0.7837 - val_loss: 0.5272 - val_accuracy: 0.8402
Epoch 3/15
47/47 [==============================] - 1s 14ms/step - loss: 0.4436 - accuracy: 0.8715 - val_loss: 0.4313 - val_accuracy: 0.8688
Epoch 4/15
47/47 [==============================] - 1s 14ms/step - loss: 0.3516 - accuracy: 0.8964 - val_loss: 0.3832 - val_accuracy: 0.8830
Epoch 5/15
47/47 [==============================] - 1s 13ms/step - loss: 0.2915 - accuracy: 0.9115 - val_loss: 0.3636 - val_accuracy: 0.8916
Epoch 6/15
47/47 [==============================] - 1s 14ms/step - loss: 0.2682 - accuracy: 0.9176 - val_loss: 0.3457 - val_accuracy: 0.8984
Epoch 7/15
47/47 [==============================] - 1s 14ms/step - loss: 0.2490 - accuracy: 0.9321 - val_loss: 0.3307 - val_accuracy: 0.9025
Epoch 8/15
47

In [ ]:
ntest=10000 #число тестируемых цифр
lotx = xTest[range(ntest)]
lotxExp=np.expand_dims(lotx,axis=0)
lotpredict=mymodel.predict(lotxExp)
lotpredict2=mymodel2.predict(lotxExp)
lotpredict3=mymodel3.predict(lotxExp)

print('доля неугаданных картинок в тесте при проверочной выборке=0.1: ',np.where(ytestOrg[range(ntest)]!=
               np.argmax(lotpredict,axis=2))[1].size/ntest) # доля неугаданных
print('доля неугаданных картинок в тесте  при проверочной выборке=0.5: ',np.where(ytestOrg[range(ntest)]!=
               np.argmax(lotpredict2,axis=2))[1].size/ntest)
print('доля неугаданных картинок в тесте  при проверочной выборке=0.9: ',np.where(ytestOrg[range(ntest)]!=
               np.argmax(lotpredict3,axis=2))[1].size/ntest)               

доля неугаданных при проверочной выборке=0.1:  0.0431
доля неугаданных при проверочной выборке=0.5:  0.0514
доля неугаданных при проверочной выборке=0.9:  0.0843


Выводы.

При повышении доли проверочной выборки с 0.1 до 0.9 точность на ней падает с 96% до 91% при точности на обучающей выборке во всех случаях примерно 96%. И также падает точность на тестовой выборке. Потому что падает доля обучающей выборки, которая должна быть как можно больше. Поэтому выборка 0.1 наиболее оптимальна из этих трех, т.к. она максимальна.

Кроме того я проверил, что понижение доли проверочной выборки еще меньше 0.1  хотя и увеличивает точность проверочной выборки, но на тестовой выборке точность остается те же 96%, т.к. практически не меняется величина обучающей выборки, при этом проверочная выборка становится слишком мала для адекватной оценки модели.

Значит проверочная выборка 0.1 близка к оптимальной из всего дапазона [0;1]
